librad/src/net/peer.rs

```diff
@@ -549,47 +549,41 @@ where
                         return PutResult::Error;
                     },
                 };
-                let res = match has.rev {
-                    // TODO: may need to fetch eagerly if we tracked while offline (#141)
-                    Some(Rev::Git(head)) if is_tracked => {
-                        let res = {
-                            let this = self.clone();
-                            this.git_fetch(provider, urn, head)
-                                .instrument(span.clone())
-                                .await
-                        };
-
-                        match res {
-                            Ok(()) => {
-                                if self.ask(has.clone()).instrument(span.clone()).await {
-                                    PutResult::Applied(has.clone())
-                                } else {
-                                    span.in_scope(|| {
-                                        tracing::warn!(
-                                            provider = %provider,
-                                            has.origin = ?has.origin,
-                                            has.urn = %has.urn,
-                                            "Provider announced non-existent rev"
-                                        );
-                                        PutResult::Stale
-                                    })
-                                }
+                let res = if is_tracked {
+                    let res = self
+                        .git_fetch(provider, urn, has.rev.as_ref().map(|Rev::Git(head)| *head))
+                        .instrument(span.clone())
+                        .await;
+
+                    match res {
+                        Ok(()) => {
+                            if self.ask(has.clone()).instrument(span.clone()).await {
+                                PutResult::Applied(has.clone())
+                            } else {
+                                span.in_scope(|| {
+                                    tracing::warn!(
+                                        provider = %provider,
+                                        has.origin = ?has.origin,
+                                        has.urn = %has.urn,
+                                        "Provider announced non-existent rev"
+                                    );
+                                    PutResult::Stale
```

Comment:
@kim Shouldn't this really be an error? We did a successful fetch but for some reason, it wasn't in our store.

Reply:
Yeaaaahh... I'm not sure. `Error` makes an effort to request retransmission. So if a peer announces things it doesn't actually have, it can cause a lot of traffic. It is probably better to just stop disseminating gossip here.

Reply:
Ok, makes sense :+1: